File: TinkerRocketAndroid/core/session/src/main/kotlin/com/tinkerbug/tinkerrocket/session/PreflightStore.kt
package com.tinkerbug.tinkerrocket.session

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.io.File
import java.util.UUID

/**
 * Pre-flight checklist store — port of iOS PreflightStore.swift.
 *
 * Persistence mirrors [RocketProfileStore]: JSON files under [directory]
 * (master.json + one <profile-uuid>.json per configured rocket), atomic
 * writes, skip-corrupt on load — a bad file loses one rocket's diff, never
 * the master or the set.  A config whose profile was since deleted is
 * inert; deleteConfig() is called from the profile-delete UI.
 *
 * Single-writer: call mutators from the fleet dispatcher.
 */
public class PreflightStore(
    private val directory: File,
    private val nowMs: () -> Long = System::currentTimeMillis,
) {
    private val _master = MutableStateFlow(PreflightMaster(updatedAtMs = 0L))
    public val master: StateFlow<PreflightMaster> = _master.asStateFlow()

    private val _configs = MutableStateFlow<Map<UUID, PreflightRocketConfig>>(emptyMap())
    public val configs: StateFlow<Map<UUID, PreflightRocketConfig>> = _configs.asStateFlow()

    init {
        load()
    }

    // ── Master CRUD ──────────────────────────────────────────────────────

    public fun addMasterItem(item: PreflightItem): PreflightItem {
        mutateMaster { it.copy(items = it.items + item) }
        return item
    }

    public fun updateMasterItem(id: UUID, mutate: (PreflightItem) -> PreflightItem) {
        mutateMaster { m ->
            m.copy(items = m.items.map { if (it.id == id) mutate(it).copy(id = id) else it })
        }
    }

    /**
     * Remove a master step everywhere: from the master, from every rocket's
     * exclusion list, and from every rocket's checked state.
     */
    public fun deleteMasterItem(id: UUID) {
        mutateMaster { m -> m.copy(items = m.items.filterNot { it.id == id }) }
        val key = id.toString().uppercase()
        for (profileId in _configs.value.keys) {
            mutateConfig(profileId) { c ->
                c.copy(
                    disabledMasterIds = c.disabledMasterIds.filterNot { it == id },
                    checked = c.checked - key,
                )
            }
        }
    }

    /** Move the step at [fromIndex] to [toIndex]; out-of-range is a no-op. */
    public fun moveMasterItem(fromIndex: Int, toIndex: Int) {
        mutateMaster { m -> m.copy(items = moved(m.items, fromIndex, toIndex)) }
    }

    /** Auto kinds already in the master — the add menu greys these out. */
    public fun masterAutoKinds(): Set<PreflightItemKind> =
        _master.value.items.map { it.kind }.filter { it.isAuto }.toSet()

    // ── Per-rocket config ────────────────────────────────────────────────

    /** Stored config for a rocket, or null if never customized. */
    public fun config(profileId: UUID): PreflightRocketConfig? = _configs.value[profileId]

    /** This rocket's effective checklist (master minus exclusions + extras). */
    public fun effectiveItems(profileId: UUID): List<PreflightItem> =
        PreflightChecklist.effectiveItems(_master.value, _configs.value[profileId])

    public fun setMasterItem(itemId: UUID, enabled: Boolean, profileId: UUID) {
        mutateConfig(profileId) { c ->
            val without = c.disabledMasterIds.filterNot { it == itemId }
            if (enabled) {
                c.copy(disabledMasterIds = without)
            } else {
                // A re-included step must come back UNCHECKED — the old
                // check is stale evidence.
                c.copy(
                    disabledMasterIds = without + itemId,
                    checked = c.checked - itemId.toString().uppercase(),
                )
            }
        }
    }

    public fun addExtraItem(item: PreflightItem, profileId: UUID): PreflightItem {
        mutateConfig(profileId) { it.copy(extraItems = it.extraItems + item) }
        return item
    }

    public fun updateExtraItem(
        itemId: UUID,
        profileId: UUID,
        mutate: (PreflightItem) -> PreflightItem,
    ) {
        mutateConfig(profileId) { c ->
            c.copy(
                extraItems = c.extraItems.map {
                    if (it.id == itemId) mutate(it).copy(id = itemId) else it
                },
            )
        }
    }

    public fun deleteExtraItem(itemId: UUID, profileId: UUID) {
        mutateConfig(profileId) { c ->
            c.copy(
                extraItems = c.extraItems.filterNot { it.id == itemId },
                checked = c.checked - itemId.toString().uppercase(),
            )
        }
    }

    /** Move a rocket's extra step at [fromIndex] to [toIndex]; out-of-range no-ops. */
    public fun moveExtraItem(profileId: UUID, fromIndex: Int, toIndex: Int) {
        mutateConfig(profileId) { c ->
            c.copy(extraItems = moved(c.extraItems, fromIndex, toIndex))
        }
    }

    /** Drop a rocket's whole config (called when its profile is deleted). */
    public fun deleteConfig(profileId: UUID) {
        if (profileId !in _configs.value) return
        _configs.value = _configs.value - profileId
        fileFor(profileId).delete()
    }

    // ── Run state (manual checks) ────────────────────────────────────────

    public fun setChecked(itemId: UUID, checked: Boolean, profileId: UUID) {
        val key = itemId.toString().uppercase()
        mutateConfig(profileId) { c ->
            if (checked) {
                c.copy(checked = c.checked + (key to nowMs()))
            } else {
                c.copy(checked = c.checked - key)
            }
        }
    }

    public fun isChecked(itemId: UUID, profileId: UUID): Boolean =
        _configs.value[profileId]?.isChecked(itemId) ?: false

    /** Clear every manual check for the next flight. */
    public fun resetRun(profileId: UUID) {
        mutateConfig(profileId) { it.copy(checked = emptyMap()) }
    }

    private fun moved(
        items: List<PreflightItem>,
        fromIndex: Int,
        toIndex: Int,
    ): List<PreflightItem> {
        if (fromIndex !in items.indices || toIndex !in items.indices) return items
        val mutable = items.toMutableList()
        mutable.add(toIndex, mutable.removeAt(fromIndex))
        return mutable
    }

    // ── Disk ─────────────────────────────────────────────────────────────

    private fun masterFile() = File(directory, "master.json")

    private fun fileFor(profileId: UUID) =
        File(directory, "${profileId.toString().uppercase()}.json")

    private fun mutateMaster(mutate: (PreflightMaster) -> PreflightMaster) {
        val updated = mutate(_master.value).copy(updatedAtMs = nowMs())
        _master.value = updated
        write(masterFile(), PreflightCodec.encodeMaster(updated))
    }

    /** All config edits funnel through here — get-or-create, stamp, persist. */
    private fun mutateConfig(
        profileId: UUID,
        mutate: (PreflightRocketConfig) -> PreflightRocketConfig,
    ) {
        val current = _configs.value[profileId]
            ?: PreflightRocketConfig(profileId = profileId, updatedAtMs = nowMs())
        val updated = mutate(current).copy(profileId = profileId, updatedAtMs = nowMs())
        _configs.value = _configs.value + (profileId to updated)
        write(fileFor(profileId), PreflightCodec.encodeConfig(updated))
    }

    private fun load() {
        directory.mkdirs()
        val masterText = masterFile().takeIf { it.exists() }
            ?.let { runCatching { it.readText() }.getOrNull() }
        _master.value = masterText?.let { PreflightCodec.decodeMaster(it, nowMs()) }
            ?: PreflightMaster(updatedAtMs = nowMs())

        val loaded = directory.listFiles { f -> f.extension == "json" && f.name != "master.json" }
            .orEmpty()
            .mapNotNull { f ->
                // Skip-corrupt: one bad file loses one rocket's diff, never the set.
                runCatching { PreflightCodec.decodeConfig(f.readText(), nowMs()) }.getOrNull()
            }
            .associateBy { it.profileId }
        _configs.value = loaded
    }

    private fun write(file: File, text: String) {
        directory.mkdirs()
        val tmp = File(directory, "${file.name}.tmp")
        tmp.writeText(text)
        check(tmp.renameTo(file) || file.let { tmp.copyTo(it, overwrite = true); tmp.delete(); true })
    }
}

File: TinkerRocketApp/TinkerRocketApp/Models/PreflightStore.swift
//
//  PreflightStore.swift
//  TinkerRocketApp
//
//  Owns the master pre-flight checklist and the per-rocket configs.
//
//  Persistence mirrors RocketProfileStore: JSON files under Application
//  Support (PreflightChecklist/master.json + one <profile-uuid>.json per
//  configured rocket), so a corrupt file loses one rocket's diff, never
//  the master or the set.  The directory is injectable for unit tests.
//
//  A config file whose profile was since deleted is inert (nothing lists
//  it) — deleteConfig(_:) is called from the profile-delete UI, and any
//  orphan that slips through is a few hundred bytes of harmless JSON.
//

import Foundation
import Combine

final class PreflightStore: ObservableObject {
    // Same swiftlang/swift#87316 workaround as RocketProfileStore: no
    // deinit-time logic, so skip the MainActor executor hop that crashes
    // synchronous XCTest teardown.
    nonisolated deinit {}

    @Published private(set) var master = PreflightMaster()
    @Published private(set) var configs: [UUID: PreflightRocketConfig] = [:]

    private let dir: URL

    init(directory: URL? = nil) {
        if let directory {
            self.dir = directory
        } else {
            let base = FileManager.default.urls(for: .applicationSupportDirectory,
                                                in: .userDomainMask)[0]
            self.dir = base.appendingPathComponent("PreflightChecklist", isDirectory: true)
        }
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        load()
    }

    // MARK: - Master CRUD

    /// Append a step to the master list.
    @discardableResult
    func addMasterItem(_ item: PreflightItem) -> PreflightItem {
        mutateMaster { $0.items.append(item) }
        return item
    }

    func updateMasterItem(_ id: UUID, _ mutate: (inout PreflightItem) -> Void) {
        mutateMaster { m in
            guard let idx = m.items.firstIndex(where: { $0.id == id }) else { return }
            mutate(&m.items[idx])
        }
    }

    /// Remove a master step everywhere: from the master, from every
    /// rocket's exclusion list, and from every rocket's checked state.
    func deleteMasterItem(_ id: UUID) {
        mutateMaster { $0.items.removeAll { $0.id == id } }
        for profileId in configs.keys {
            mutateConfig(profileId) { c in
                c.disabledMasterIds.removeAll { $0 == id }
                c.checked.removeValue(forKey: id.uuidString)
            }
        }
    }

    func moveMasterItems(fromOffsets: IndexSet, toOffset: Int) {
        mutateMaster { Self.move(&$0.items, fromOffsets: fromOffsets, toOffset: toOffset) }
    }

    /// Auto kinds already present in the master — the add menu greys these
    /// out (two "camera recording" steps verify the same bit twice).
    var masterAutoKinds: Set<PreflightItemKind> {
        Set(master.items.map(\.kind).filter(\.isAuto))
    }

    // MARK: - Per-rocket config

    /// The stored config for a rocket, or nil if it has never been
    /// customized (= full master list, nothing checked).
    func config(for profileId: UUID) -> PreflightRocketConfig? {
        configs[profileId]
    }

    /// This rocket's effective checklist (master minus exclusions + extras).
    func effectiveItems(for profileId: UUID) -> [PreflightItem] {
        PreflightChecklist.effectiveItems(master: master, config: configs[profileId])
    }

    func setMasterItem(_ itemId: UUID, enabled: Bool, for profileId: UUID) {
        mutateConfig(profileId) { c in
            c.disabledMasterIds.removeAll { $0 == itemId }
            if !enabled {
                c.disabledMasterIds.append(itemId)
                c.checked.removeValue(forKey: itemId.uuidString)
            }
        }
    }

    @discardableResult
    func addExtraItem(_ item: PreflightItem, for profileId: UUID) -> PreflightItem {
        mutateConfig(profileId) { $0.extraItems.append(item) }
        return item
    }

    func updateExtraItem(_ itemId: UUID, for profileId: UUID,
                         _ mutate: (inout PreflightItem) -> Void) {
        mutateConfig(profileId) { c in
            guard let idx = c.extraItems.firstIndex(where: { $0.id == itemId }) else { return }
            mutate(&c.extraItems[idx])
        }
    }

    func deleteExtraItem(_ itemId: UUID, for profileId: UUID) {
        mutateConfig(profileId) { c in
            c.extraItems.removeAll { $0.id == itemId }
            c.checked.removeValue(forKey: itemId.uuidString)
        }
    }

    func moveExtraItems(fromOffsets: IndexSet, toOffset: Int, for profileId: UUID) {
        mutateConfig(profileId) { Self.move(&$0.extraItems, fromOffsets: fromOffsets, toOffset: toOffset) }
    }

    /// SwiftUI's `Array.move(fromOffsets:toOffset:)` semantics without the
    /// SwiftUI import (this is a model file): remove the offsets, then
    /// insert the block at the destination adjusted for the removals above it.
    private static func move(_ items: inout [PreflightItem],
                             fromOffsets: IndexSet, toOffset: Int) {
        let moving = fromOffsets.sorted(by: >).map { items.remove(at: $0) }.reversed()
        let dest = toOffset - fromOffsets.count(where: { $0 < toOffset })
        items.insert(contentsOf: moving, at: dest)
    }

    /// Drop a rocket's whole config (called when its profile is deleted).
    func deleteConfig(for profileId: UUID) {
        guard configs[profileId] != nil else { return }
        configs.removeValue(forKey: profileId)
        try? FileManager.default.removeItem(at: configURL(profileId))
    }

    // MARK: - Run state (manual checks)

    func setChecked(_ itemId: UUID, checked: Bool, for profileId: UUID) {
        mutateConfig(profileId) { c in
            if checked {
                c.checked[itemId.uuidString] = Date()
            } else {
                c.checked.removeValue(forKey: itemId.uuidString)
            }
        }
    }

    func isChecked(_ itemId: UUID, for profileId: UUID) -> Bool {
        configs[profileId]?.isChecked(itemId) ?? false
    }

    /// Clear every manual check for the next flight.
    func resetRun(for profileId: UUID) {
        mutateConfig(profileId) { $0.checked = [:] }
    }

    // MARK: - Persistence

    private var masterURL: URL { dir.appendingPathComponent("master.json") }

    private func configURL(_ profileId: UUID) -> URL {
        dir.appendingPathComponent("\(profileId.uuidString).json")
    }

    private func mutateMaster(_ mutate: (inout PreflightMaster) -> Void) {
        mutate(&master)
        master.updatedAt = Date()
        write(master, to: masterURL)
    }

    /// All config edits funnel through here — get-or-create, bump
    /// `updatedAt`, persist — so none of the three can be forgotten.
    private func mutateConfig(_ profileId: UUID,
                              _ mutate: (inout PreflightRocketConfig) -> Void) {
        var config = configs[profileId] ?? PreflightRocketConfig(profileId: profileId)
        mutate(&config)
        config.updatedAt = Date()
        configs[profileId] = config
        write(config, to: configURL(profileId))
    }

    private func load() {
        let decoder = JSONDecoder()
        if let data = try? Data(contentsOf: masterURL),
           let m = try? decoder.decode(PreflightMaster.self, from: data) {
            master = m
        }

        let urls = (try? FileManager.default.contentsOfDirectory(
            at: dir, includingPropertiesForKeys: nil)) ?? []
        var loaded: [UUID: PreflightRocketConfig] = [:]
        for url in urls where url.pathExtension == "json" && url != masterURL {
            guard let data = try? Data(contentsOf: url),
                  let c = try? decoder.decode(PreflightRocketConfig.self, from: data)
            else { continue }   // skip corrupt/foreign files — don't fail the set
            loaded[c.profileId] = c
        }
        configs = loaded
    }

    private func write<T: Encodable>(_ value: T, to url: URL) {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let data = try? encoder.encode(value) else { return }
        try? data.write(to: url, options: .atomic)
    }
}
